src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderCng.cs

```diff
@@ -125,21 +136,27 @@ public sealed override void Dispose(bool disposing)
 
         public override void Reset()
         {
+            // Reset does not need to use ConcurrencyBlock. It either no-ops, or creates an entirely new handle, exchanges
+            // them, and disposes of the old handle. We don't need to block concurrency on the Dispose because SafeHandle
+            // does that.
             if (_reusable && !_running)
                 return;
 
-            DestroyHash();
-
             BCryptCreateHashFlags flags = _reusable ?
                 BCryptCreateHashFlags.BCRYPT_HASH_REUSABLE_FLAG :
                 BCryptCreateHashFlags.None;
 
             SafeBCryptHashHandle hHash;
             NTSTATUS ntStatus = Interop.BCrypt.BCryptCreateHash(_hAlgorithm, out hHash, IntPtr.Zero, 0, _key, _key == null ? 0 : _key.Length, flags);
+
             if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+            {
+                hHash.Dispose();
                 throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+            }
 
-            _hHash = hHash;
+            SafeBCryptHashHandle? previousHash = Interlocked.Exchange(ref _hHash, hHash);
```

Comment:
Assuming this made it past the `_running` guard, this

1. Disposed the current handle in DestroyHash
2. Nulled out `_hHash`
3. Created a new hash
4. Assigned `_hHash` to the new one.

If another thread attempted to use that hash during 2 or 3, then some asserts would trip because we never expected `_hHash` to be null.

Now, we

1. Create a new hash.
2. Exchange old and new
3. Dispose of old.

This ensures we don't have a period of time where _hHash can be null and trip asserts during test runs.